swim/suspicion.go

```diff
@@ -33,62 +33,85 @@ type suspect interface {
 	incarnation() int64
 }
 
+type timer struct {
+	*time.Timer
+	// at which state the timer was started
+	state string
+}
+
 // Suspicion handles the suspicion sub-protocol of the SWIM protocol
 type suspicion struct {
 	sync.Mutex
 
 	node *Node
 
-	timeout time.Duration
-	timers  map[string]*time.Timer
-	enabled bool
-	logger  log.Logger
+	suspectTimeout time.Duration
+	timers         map[string]*timer
+	enabled        bool
+	logger         log.Logger
 }
 
 // newSuspicion returns a new suspicion SWIM sub-protocol with the given timeout
-func newSuspicion(n *Node, timeout time.Duration) *suspicion {
+func newSuspicion(n *Node, suspectTimeout time.Duration) *suspicion {
 	suspicion := &suspicion{
-		node:    n,
-		timeout: timeout,
-		timers:  make(map[string]*time.Timer),
-		enabled: true,
-		logger:  logging.Logger("suspicion").WithField("local", n.Address()),
+		node:           n,
+		suspectTimeout: suspectTimeout,
+		timers:         make(map[string]*timer),
+		enabled:        true,
+		logger:         logging.Logger("suspicion").WithField("local", n.Address()),
 	}
 
 	return suspicion
 }
 
-func (s *suspicion) Start(suspect suspect) {
-	s.withLock(func() {
-		if !s.enabled {
-			s.logger.Warn("cannot start suspect period while disabled")
-			return
-		}
+func (s *suspicion) StartSuspect(suspect suspect) {
+	s.Lock()
+	s.start(suspect, Suspect)
+	s.Unlock()
+}
 
-		if s.node.Address() == suspect.address() {
-			s.logger.Warn("cannot start suspect period for local member")
-			return
-		}
+func (s *suspicion) start(suspect suspect, state string) {
+	if !s.enabled {
+		s.logger.Warn("cannot start suspect period while disabled")
+		return
+	}
 
-		if _, ok := s.timers[suspect.address()]; ok {
+	if s.node.Address() == suspect.address() {
+		s.logger.Warn("cannot start suspect period for local member")
+		return
+	}
+
+	if t, ok := s.timers[suspect.address()]; ok {
+		if t.state == state {
 			s.logger.Warn("redundant call to start suspect ignored")
 			return
 		}
+		// try cancel the timer for a different state transition
+		t.Stop()
+	}
 
-		s.timers[suspect.address()] = time.AfterFunc(s.timeout, func() {
-			s.logger.WithField("faulty", suspect.address()).Info("member declared faulty")
+	t := time.AfterFunc(s.suspectTimeout, func() {
+		s.logger.WithField(state, suspect.address()).Infof("member declared %s", state)
+		switch state {
+		case Suspect:
 			s.node.memberlist.MakeFaulty(suspect.address(), suspect.incarnation())
-		})
-
-		s.logger.WithField("suspect", suspect.address()).Debug("started member suspect period")
+		default:
+			panic("unknown state")
+		}
 	})
+	s.timers[suspect.address()] = &timer{
+		Timer: t,
+		state: state,
+	}
+
+	s.logger.WithField(state, suspect.address()).Debugf("started member %s period", state)
```

Comment:
A bit more information in the log message might be nice, something like: `Started state timer for member`.

And is it intentional that you encode the state as a field name in the `WithField`? it might be easier for machines to parse if you add specific fields for `address` and `state`